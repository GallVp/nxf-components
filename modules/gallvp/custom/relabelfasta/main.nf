process CUSTOM_RELABELFASTA {
    tag "$meta.id"
    label 'process_single'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/biopython:1.75':
        'biocontainers/biopython:1.75' }"

    input:
    tuple val(meta), path(fasta)
    path(labels)

    output:
    tuple val(meta), path("*.fasta")    , emit: fasta
    path "versions.yml"                 , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args    = task.ext.args ?: ''
    prefix      = task.ext.prefix ?: "${meta.id}"
    if ( "${prefix}.fasta" == "$fasta" ) error "Input and output names are the same, use \"task.ext.prefix\" to disambiguate!"
    template 'relabelfasta.py'

    stub:
    def args    = task.ext.args ?: ''
    def prefix  = task.ext.prefix ?: "${meta.id}"
    if ( "${prefix}.fasta" == "$fasta" ) error "Input and output names are the same, use \"task.ext.prefix\" to disambiguate!"
    """
    touch ${prefix}.fasta

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        python: \$(python --version | cut -d' ' -f2)
        biopython: \$(pip list | grep "biopython" | cut -d' ' -f3)
    END_VERSIONS
    """
}
