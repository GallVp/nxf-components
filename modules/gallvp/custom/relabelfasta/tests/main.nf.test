nextflow_process {

    name "Test Process CUSTOM_RELABELFASTA"
    script "../main.nf"
    process "CUSTOM_RELABELFASTA"

    tag "modules"
    tag "modules_gallvp"
    tag "custom"
    tag "custom/relabelfasta"

    test("sarscov2 - fasta") {
        when {
            process {
                """
                input[0] = [
                    [ id:'test' ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                ]
                input[1] = Channel.of('MT192765.1\tChr1').collectFile(name: 'id_map.tsv', newLine: true)
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }

            )
        }

    }

    test("sarscov2 - fasta - repeated labels") {
        when {
            process {
                """
                input[0] = [
                    [ id:'test' ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                ]
                input[1] = Channel.of('MT192765.1\\tChr1\\nMT19276\\tChr1').collectFile(name: 'id_map.tsv', newLine: true)
                """
            }
        }

        then {
            assertAll(
                { assert ! process.success },
                { assert process.stdout.toString().contains('Each sequence label should be unique') }

            )
        }

    }

    test("sarscov2 - fasta - repeated ids") {
        when {
            process {
                """
                input[0] = [
                    [ id:'test' ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                ]
                input[1] = Channel.of('MT192765.1\\tChr1\\nMT192765.1\\tChr1').collectFile(name: 'id_map.tsv', newLine: true)
                """
            }
        }

        then {
            assertAll(
                { assert ! process.success },
                { assert process.stdout.toString().contains('Each sequence ID should be unique') }

            )
        }

    }

    test("sarscov2 - fasta - malformed") {
        when {
            process {
                """
                input[0] = [
                    [ id:'test' ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                ]
                input[1] = Channel.of('MT192765.1\\tChr1\\nMT192765.1').collectFile(name: 'id_map.tsv', newLine: true)
                """
            }
        }

        then {
            assertAll(
                { assert ! process.success },
                { assert process.stdout.toString().contains('should be a two column TSV file') }

            )
        }

    }

    test("sarscov2 - fasta - missing ids") {
        when {
            process {
                """
                input[0] = [
                    [ id:'test' ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                ]
                input[1] = Channel.of('MT192765.1\\tChr1\\nMT192765.2\\tChr2').collectFile(name: 'id_map.tsv', newLine: true)
                """
            }
        }

        then {
            assertAll(
                { assert ! process.success },
                { assert process.stdout.toString().contains('missing from the fasta') }

            )
        }

    }

    test("sarscov2 - fasta - stub") {

        options '-stub'

        when {
            process {
                """
                input[0] = [
                    [ id:'test' ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                ]
                input[1] = Channel.of('MT192765.1\tChr1').collectFile(name: 'id_map.tsv', newLine: true)
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }

            )
        }

    }

}
