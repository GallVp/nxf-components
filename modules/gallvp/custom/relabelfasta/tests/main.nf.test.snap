{
    "sarscov2 - fasta - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.fasta:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    "versions.yml:md5,d2052e0039190dca1b50700028ef14c3"
                ],
                "fasta": [
                    [
                        {
                            "id": "test"
                        },
                        "test.fasta:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,d2052e0039190dca1b50700028ef14c3"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.1"
        },
        "timestamp": "2024-05-25T13:40:44.896721"
    },
    "sarscov2 - fasta": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.fasta:md5,393572534251b71ec74fbe3b1a31a709"
                    ]
                ],
                "1": [
                    "versions.yml:md5,d2052e0039190dca1b50700028ef14c3"
                ],
                "fasta": [
                    [
                        {
                            "id": "test"
                        },
                        "test.fasta:md5,393572534251b71ec74fbe3b1a31a709"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,d2052e0039190dca1b50700028ef14c3"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.1"
        },
        "timestamp": "2024-05-25T13:38:20.323976"
    }
}